upup/pkg/fi/cloudup/urls.go

```diff
@@ -17,74 +17,167 @@ limitations under the License.
 package cloudup
 
 import (
+	"fmt"
+	"net/url"
 	"os"
-	"strings"
+
+	"path"
 
 	"github.com/golang/glog"
 	"k8s.io/kops"
+	"k8s.io/kops/pkg/assets"
+	"k8s.io/kops/util/pkg/hashing"
 )
 
-// baseUrl caches the BaseUrl value
-var baseUrl string
+const kopsUrl = "https://kubeupv2.s3.amazonaws.com/kops/"
+
+// kopsBaseUrl caches the base URL for kops binaries
+var kopsBaseUrl *url.URL
+
+// nodeUpLocation caches the nodeUpLocation url
+var nodeUpLocation *url.URL
+
+// nodeUpHash caches the hash for nodeup
+var nodeUpHash *hashing.Hash
+
+// protokubeLocation caches the protokubeLocation url
+var protokubeLocation *url.URL
+
+// protokubeHash caches the hash for protokube
+var protokubeHash *hashing.Hash
 
 // BaseUrl returns the base url for the distribution of kops - in particular for nodeup & docker images
-func BaseUrl() string {
-	if baseUrl != "" {
-		// Avoid repeated logging
-		return baseUrl
+func BaseUrl() (*url.URL, error) {
+	// returning cached value
```

Comment:
Can you keep the  "avoid repeated logging" comment - that's why it's important to cache this - not for speed.